src/Microsoft.AspNetCore.Server.Kestrel/ListenOptions.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+using Microsoft.AspNetCore.Server.Kestrel.Adapter;
+
+namespace Microsoft.AspNetCore.Server.Kestrel
+{
+    /// <summary>
+    /// Describes either an <see cref="IPEndPoint"/>, Unix domain socket path, or a file descriptor for an already open
+    /// socket that Kestrel should bind to or open.
+    /// </summary>
+    public class ListenOptions
+    {
+        private bool _noDelay = true;
+
+        internal ListenOptions(IPEndPoint endPoint)
+        {
+            Type = ListenType.IPEndPoint;
+            IPEndPoint = endPoint;
+        }
+
+        internal ListenOptions(string socketPath)
+        {
+            Type = ListenType.SocketPath;
+            SocketPath = socketPath;
+        }
+
+        internal ListenOptions(long fileDescriptor)
+        {
+            Type = ListenType.FileDescriptor;
+            FileDescriptor = fileDescriptor;
+        }
+
+        /// <summary>
+        /// The type of interface being described: either an <see cref="IPEndPoint"/>, Unix domain socket path, or a file descriptor.
+        /// </summary>
+        public ListenType Type { get; }
+
+        // IPEndPoint is mutable so port 0 can be updated to the bound port.
+        public IPEndPoint IPEndPoint { get; internal set; }
+        public string SocketPath { get; }
+        public long FileDescriptor { get; }
+
+        /// <summary>
+        /// Enables and <see cref="IConnectionAdapter"/> to resolve and use services registered by the application during startup.
+        /// Only set if accessed from the callback of a <see cref="KestrelServerOptions"/> Listen* method.
+        /// </summary>
+        public KestrelServerOptions KestrelServerOptions { get; internal set; }
+
+        /// <summary>
+        /// Set to false to enable Nagle's algorithm for all connections.
+        /// </summary>
+        /// <remarks>
+        /// Defaults to true.
+        /// </remarks>
+        public bool NoDelay
```

Comment:
Even though we call uv_tcp_nodelay on the listen socket, I don't think it has any effect there (it *might* get inherited but that's pretty moot). I think the only important thing is configuring nodelay for the accepted sockets.

I certainly wouldn't split the configuration. We could remove the call uv_tcp_nodelay for the listen socket, but I would do that as part of another change and make absolutely sure that it doesn't affect perf.